genesis/streaming.py

```diff
@@ -14,350 +14,414 @@
+            unit=" alerts",
+            mininterval=0.5,
+            smoothing=0.5,
+            miniters=0,
+        )
+
+        nread = 0
+        for rec, meta in self._filtered_stream(
+            mapper=mapper, filter=filter, timeout=timeout
+        ):
+            yield (rec, meta) if self._metadata else rec
+
+            t.update()
+            nread += 1
+
+            if nread == limit:
+                break
+
+        t.close()
+
+    # returns a generator for the user-mapped messages using the filter function,
+    # possibly executed on ncores and up to maxread values
+    def __call__(
+        self, filter=None, pool=None, progress=False, timeout=None, limit=None
+    ):
+        if pool:
+            mapper = lambda fun, vec: pool.imap(fun, vec, chunksize=100)
+        else:
+            mapper = map
+
+        yield from self._stream(
+            filter, mapper=mapper, progress=progress, timeout=timeout, limit=limit
+        )
+
+    def __iter__(self):
+        return self.__call__()
+
+    # producer functionality
+    def write(self, msg):
+        packet = self._serialize(msg)
+        self.p.produce(self.topics[0], packet)
+
+    def flush(self):
+        return self.p.flush()
+
 
 @contextmanager
 def Pool(*args, **kwarg):
-	original_sigint_handler = signal.signal(signal.SIGINT, signal.SIG_IGN)
-	p = MPPool(*args, **kwarg)
-	signal.signal(signal.SIGINT, original_sigint_handler)
-	try:
-		yield p
-	finally:
-		p.close()
-
-if __name__ == "__main__":
-	def my_filter(msg):
-		#return msg
-		return None if msg.candidate.ssnamenr == 'null' else msg
-
-	try:
-		from datetime import datetime
-		with Pool(5) as workers:
-			with AlertBroker("kafka://broker0.do.alerts.wtf/test6", start_at="earliest") as stream:
-				for nread, (idx, rec) in enumerate(stream(filter=my_filter, pool=workers, progress=True, timeout=10), start=1):
-
-					## do stuff
-					cd = rec.candidate
-					print(f"[{datetime.now()}] {nread}/{idx}:", cd.jd, cd.ssdistnr, cd.ssnamenr)
+    original_sigint_handler = signal.signal(signal.SIGINT, signal.SIG_IGN)
+    p = MPPool(*args, **kwarg)
+    signal.signal(signal.SIGINT, original_sigint_handler)
+    try:
+        yield p
+    finally:
+        p.close()
 
-					stream.commit()
-				stream.commit(defer=False)
 
-		# with AlertBroker("kafka://broker0.do.alerts.wtf/test8", start_at="earliest") as stream:
-		# 	for nread, (idx, rec) in enumerate(stream(progress=True, timeout=2), start=1):
-
-		# 		## do stuff
-		# 		cd = rec.candidate
-		# 		print(f"[{datetime.now()}] {nread}/{idx}:", cd.jd, cd.ssdistnr, cd.ssnamenr)
-
-		# 		stream.commit()
-
-		# 		if nread == 10:
-		# 			break
-
-		# 	print("OUTSEDE")
-		# 	print("GOING IN")
-
-		# 	for nread, (idx, rec) in enumerate(stream(progress=True, timeout=10), start=1):
-		# 		cd = rec.candidate
-		# 		print(f"[{datetime.now()}] {nread}/{idx}:", cd.jd, cd.ssdistnr, cd.ssnamenr)
-		# 		stream.commit()
+if __name__ == "__main__":
```

Comment:
As you've probably figured out, this is all ad-hoc debugging/testing junk which should go away once a proper test suite is in place.

Reply:
Yep, I plan to do all that sort of cleanup only after I have automatic formatting and linting. That way I don't have to rewrite any of my own code, at least :)